File: electron/agent/permissions/rules.ts
/**
 * Rule-based permission rules — `.agents/settings.json`.
 *
 * Rule format: `"ToolName(argPattern)"` — e.g. `"bash(pnpm i)"`,
 * `"bash(npx:*)"`, `"edit_file(src/)"`. Tool name matches case-insensitively
 * via `'*'` / exact / `'prefix:*'`.
 *
 * argPattern matching:
 *   - Bare `"ToolName"` (no parens) = match any args
 *   - `"ToolName(prefix)"` = prefix match on the tool's primary arg
 *   - `"ToolName(prefix:*)"` = glob match — prefix up to `:*` matches anything after
 *   - `"ToolName(*suffix)"` = suffix glob match
 *   - `"ToolName(*middle*)"` = contains glob match
 *
 * Precedence: deny always rejects; allow upgrades an 'ask' decision to auto;
 * allow does NOT bypass plan-mode blocking.
 *
 * Rules persist in `{workspaceRoot}/.agents/settings.json`:
 * {
 *   "permissions": {
 *     "allow": ["bash(npm:*)", "edit_file(src/*)"],
 *     "deny": []
 *   }
 * }
 *
 * There is no separate "session" vs "project" scope — all rules are project-level
 * and persist across sessions. This matches Claude Code's behavior.
 */
import * as fs from 'fs';
import * as path from 'path';
import { minimatch } from 'minimatch';

export interface Rule {
  /** Tool-name pattern: '*' | 'bash' | 'edit:*' (case-insensitive). */
  tool: string;
  /** Pattern on the tool's primary arg; null = match any args. */
  argPattern: string | null;
}

export interface RuleSet {
  allow: Rule[];
  deny: Rule[];
}

const EMPTY: RuleSet = { allow: [], deny: [] };

/** Parse `"ToolName(argPattern)"` or `"ToolName"`. null if unparseable. */
export function parseRule(spec: string): Rule | null {
  const s = spec.trim();
  if (!s) return null;
  const m = s.match(/^([^()]+)\(([\s\S]*)\)$/);
  if (m) {
    const tool = m[1].trim();
    if (!tool) return null;
    const arg = m[2].trim();
    return { tool, argPattern: arg || null };
  }
  return { tool: s, argPattern: null };
}

function parseList(raw: unknown): Rule[] {
  if (!Array.isArray(raw)) return [];
  return raw
    .map((x) => (typeof x === 'string' ? parseRule(x) : null))
    .filter((r): r is Rule => r !== null);
}

function loadFile(filePath: string): RuleSet {
  try {
    const raw = fs.readFileSync(filePath, 'utf-8');
    const parsed = JSON.parse(raw);
    if (typeof parsed !== 'object' || !parsed) return EMPTY;
    const perms = (parsed as { permissions?: { allow?: unknown; deny?: unknown } }).permissions ?? {};
    return { allow: parseList(perms.allow), deny: parseList(perms.deny) };
  } catch {
    return EMPTY;
  }
}

/** Load project rules from `<root>/.agents/settings.json`. */
export function loadPermissionRules(workspaceRoot: string): RuleSet {
  return loadFile(path.join(workspaceRoot, '.agents', 'settings.json'));
}

/** The primary arg used for argPattern matching, per tool. */
export function primaryArg(toolName: string, args: Record<string, unknown>): string | null {
  const a = args as { command?: string; args?: string[]; url?: string; query?: string; path?: string; pattern?: string };
  switch (toolName) {
    case 'bash':
      return typeof a.command === 'string' ? a.command : null;
    case 'git':
      return Array.isArray(a.args) ? a.args.join(' ') : null;
    case 'web_fetch':
      return typeof a.url === 'string' ? a.url : null;
    case 'web_search':
      return typeof a.query === 'string' ? a.query : null;
    case 'edit_file':
    case 'multi_edit':
    case 'write_file':
    case 'notebook_edit':
    case 'read_file':
    case 'list_dir':
    case 'glob':
    case 'grep':
      return typeof a.path === 'string' ? a.path : typeof a.pattern === 'string' ? a.pattern : null;
    default:
      return null;
  }
}

function toolNameMatches(pattern: string, toolName: string): boolean {
  const p = pattern.toLowerCase();
  const t = toolName.toLowerCase();
  if (p === '*') return true;
  if (p === t) return true;
  if (p.endsWith(':*')) return t.startsWith(p.slice(0, -2));
  return false;
}

/**
 * Does an arg pattern match a value? Supports:
 *   - Plain prefix: "npm install" matches "npm install --force"
 *   - Glob suffix: "npx:*" matches "npx create-react-app" (anything after npx )
 *   - Glob path: "src/*" matches "src/components/Foo.ts"
 *   - Glob wildcard: "*" matches anything
 */
function argPatternMatches(pattern: string, value: string): boolean {
  // If the pattern contains glob chars (* or ? or [), use minimatch.
  if (/[*?\[]/.test(pattern)) {
    return minimatch(value, pattern, { dot: true });
  }
  // Plain prefix match.
  return value.startsWith(pattern);
}

/** Does a rule match a specific tool call? */
export function ruleMatches(rule: Rule, toolName: string, args: Record<string, unknown>): boolean {
  if (!toolNameMatches(rule.tool, toolName)) return false;
  if (rule.argPattern === null) return true;
  const arg = primaryArg(toolName, args);
  if (arg === null) return false;
  return argPatternMatches(rule.argPattern, arg);
}

/** Evaluate rules. 'deny' wins; else 'allow'; else null. */
export function evaluateRules(
  rules: RuleSet,
  toolName: string,
  args: Record<string, unknown>,
): 'deny' | 'allow' | null {
  for (const r of rules.deny) if (ruleMatches(r, toolName, args)) return 'deny';
  for (const r of rules.allow) if (ruleMatches(r, toolName, args)) return 'allow';
  return null;
}

/**
 * Derive a rule spec for "Always Allow" from the approved call.
 * Generates smart glob patterns:
 *   - bash: "bash(npx:*)" for npx commands, "bash(npm install)" for npm
 *   - file tools: "edit_file(src/*)" for path-based tools
 *   - bare tool name if no recognizable arg
 */
export function deriveRuleSpec(toolName: string, args: Record<string, unknown>): string {
  const arg = primaryArg(toolName, args);
  if (arg === null) return toolName;

  if (toolName === 'bash') {
    // For npx commands, use a glob: "npx package-name" → "bash(npx package-name:*)"
    if (arg.startsWith('npx ')) {
      const pkg = arg.split(/\s+/).slice(0, 2).join(' '); // "npx package-name"
      return `${toolName}(${pkg}:*)`;
    }
    // For npm/yarn/pnpm commands, keep first 2 tokens as prefix.
    if (/^(npm|yarn|pnpm|bun|deno)\s/.test(arg)) {
      const head = arg.split(/\s+/).slice(0, 2).join(' ');
      return `${toolName}(${head})`;
    }
    // Other commands: first token only.
    const head = arg.split(/\s+/)[0];
    return `${toolName}(${head})`;
  }

  // File tools: use the directory as a prefix glob.
  if (arg.includes('/')) {
    const dir = arg.substring(0, arg.lastIndexOf('/'));
    return `${toolName}(${dir}/*)`;
  }

  return `${toolName}(${arg})`;
}

// ─── In-memory cache (refreshed each turn via loadPermissionRules) ─────
// Session-scoped rules are still in-memory for the current turn (so a rule
// written mid-turn is immediately visible without re-reading the file).
// They're also persisted to .agents/settings.json by addPermissionRule.
const sessionRules = new Map<string, RuleSet>();

export function addSessionRule(sessionId: string, scope: 'allow' | 'deny', rule: Rule): void {
  const cur = sessionRules.get(sessionId) ?? { allow: [], deny: [] };
  cur[scope].push(rule);
  sessionRules.set(sessionId, cur);
}

export function getSessionRules(sessionId: string): RuleSet {
  return sessionRules.get(sessionId) ?? EMPTY;
}

export function clearSessionRules(sessionId: string): void {
  sessionRules.delete(sessionId);
}

// ─── Unified rule writer (replaces addSessionRule + addProjectRule) ────

/**
 * Add an "always allow" rule to `.agents/settings.json`. Also adds it to the
 * in-memory session rules so it takes effect immediately without a file re-read.
 *
 * The rule is derived from the approved tool call via deriveRuleSpec, producing
 * smart glob patterns (e.g. bash(npx:*) for npx commands).
 */
export function addPermissionRule(
  sessionId: string,
  workspaceRoot: string,
  toolName: string,
  args: Record<string, unknown>,
): string | null {
  const spec = deriveRuleSpec(toolName, args);
  const rule = parseRule(spec);
  if (!rule) return null;

  // Add to in-memory session rules (immediate effect this turn).
  addSessionRule(sessionId, 'allow', rule);

  // Persist to .agents/settings.json (survives across sessions).
  const file = path.join(workspaceRoot, '.agents', 'settings.json');
  let cfg: { permissions: { allow: string[]; deny: string[] } } = {
    permissions: { allow: [], deny: [] },
  };
  try {
    const raw = fs.readFileSync(file, 'utf-8');
    const parsed = JSON.parse(raw);
    if (parsed && typeof parsed === 'object') {
      cfg = parsed as typeof cfg;
      if (!cfg.permissions) cfg.permissions = { allow: [], deny: [] };
      if (!Array.isArray(cfg.permissions.allow)) cfg.permissions.allow = [];
      if (!Array.isArray(cfg.permissions.deny)) cfg.permissions.deny = [];
    }
  } catch {
    // missing/malformed — start fresh.
  }
  if (!cfg.permissions.allow.includes(spec)) {
    cfg.permissions.allow.push(spec);
  }
  try {
    fs.mkdirSync(path.dirname(file), { recursive: true });
    fs.writeFileSync(file, JSON.stringify(cfg, null, 2) + '\n', 'utf-8');
  } catch {
    // best-effort — don't fail the turn over a rule write.
  }

  return spec;
}

File: src-tauri/crates/tide-tools/src/permission/rules.rs
//! Rule-based permission rules — port of `app/core/agent/permissions/rules.ts`
//! (91ec558). Spec format: `"ToolName(argPattern)"` — bare tool name matches
//! any args; `prefix`, glob (`*`/`?`/`[`) patterns match the tool's primary
//! arg. Precedence: deny wins; allow upgrades ask→auto (never bypasses plan
//! mode). Project rules persist in `.agents/settings.json`.

use std::path::Path;

use serde_json::Value;

#[derive(Debug, Clone, PartialEq)]
pub struct Rule {
    /// Tool-name pattern: `'*'` | `'bash'` | `'edit:*'` (case-insensitive).
    pub tool: String,
    /// Pattern on the tool's primary arg; `None` = match any args.
    pub arg_pattern: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq)]
pub struct RuleSet {
    pub allow: Vec<Rule>,
    pub deny: Vec<Rule>,
}

/// Parse `"ToolName(argPattern)"` or `"ToolName"`.
pub fn parse_rule(spec: &str) -> Option<Rule> {
    let s = spec.trim();
    if s.is_empty() {
        return None;
    }
    if let Some(open) = s.find('(') {
        if s.ends_with(')') && open > 0 {
            let tool = s[..open].trim();
            // TS `^([^()]+)\(` — the tool segment itself must be paren-free.
            if tool.is_empty() || tool.contains('(') || tool.contains(')') {
                return None;
            }
            let arg = s[open + 1..s.len() - 1].trim();
            return Some(Rule {
                tool: tool.to_string(),
                arg_pattern: if arg.is_empty() { None } else { Some(arg.to_string()) },
            });
        }
        return None;
    }
    Some(Rule {
        tool: s.to_string(),
        arg_pattern: None,
    })
}

fn parse_list(raw: Option<&Value>) -> Vec<Rule> {
    raw.and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|x| x.as_str())
                .filter_map(parse_rule)
                .collect()
        })
        .unwrap_or_default()
}

fn load_file(path: &Path) -> RuleSet {
    let Ok(raw) = std::fs::read_to_string(path) else {
        return RuleSet::default();
    };
    let Ok(parsed) = serde_json::from_str::<Value>(&raw) else {
        return RuleSet::default();
    };
    let perms = parsed.get("permissions").cloned().unwrap_or(Value::Null);
    RuleSet {
        allow: parse_list(perms.get("allow")),
        deny: parse_list(perms.get("deny")),
    }
}

/// Load project rules from `<root>/.agents/settings.json`. Missing or
/// malformed files yield an empty set (the gate falls back to modes).
pub fn load_permission_rules(workspace_root: &Path) -> RuleSet {
    load_file(&workspace_root.join(".agents").join("settings.json"))
}

/// The primary arg used for argPattern matching, per tool (TS `primaryArg`).
pub fn primary_arg(tool_name: &str, args: &Value) -> Option<String> {
    let get_str = |key: &str| args.get(key).and_then(|v| v.as_str()).map(String::from);
    match tool_name {
        "bash" => get_str("command"),
        "dispatch_agent" => get_str("name"),
        "git_repo" => get_str("repo"),
        "git" => args.get("args").and_then(|v| v.as_array()).map(|arr| {
            arr.iter()
                .filter_map(|x| x.as_str())
                .collect::<Vec<_>>()
                .join(" ")
        }),
        "web_fetch" => get_str("url"),
        "web_search" => get_str("query"),
        "edit_file" | "multi_edit" | "write_file" | "notebook_edit" | "read_file"
        | "list_dir" | "glob" | "grep" => get_str("path").or_else(|| get_str("pattern")),
        _ => None,
    }
}

fn tool_name_matches(pattern: &str, tool_name: &str) -> bool {
    let p = pattern.to_lowercase();
    let t = tool_name.to_lowercase();
    if p == "*" || p == t {
        return true;
    }
    if let Some(prefix) = p.strip_suffix(":*") {
        return t.starts_with(prefix);
    }
    false
}

/// Does an arg pattern match a value? Glob chars (`* ? [`) get wildcard
/// matching (`*`/`?` stay within a path segment, minimatch-style); other
/// patterns are plain prefix matches.
fn arg_pattern_matches(pattern: &str, value: &str) -> bool {
    if pattern.contains(['*', '?', '[']) {
        return wildcard_match(value, pattern);
    }
    value.starts_with(pattern)
}

/// minimatch-subset wildcard matcher (the TS used minimatch with `dot: true`).
fn wildcard_match(value: &str, pattern: &str) -> bool {
    let v: Vec<char> = value.chars().collect();
    let p: Vec<char> = pattern.chars().collect();
    wc(&v, 0, &p, 0)
}

fn wc(v: &[char], mut vi: usize, p: &[char], mut pi: usize) -> bool {
    while pi < p.len() {
        match p[pi] {
            '*' => {
                // Collapse consecutive stars; `*` does not cross '/'.
                while pi < p.len() && p[pi] == '*' {
                    pi += 1;
                }
                if pi == p.len() {
                    return !v[vi..].contains(&'/');
                }
                for next in vi..=v.len() {
                    if v.get(next) == Some(&'/') {
                        break;
                    }
                    if wc(v, next, p, pi) {
                        return true;
                    }
                }
                return false;
            }
            '?' => {
                if vi >= v.len() || v[vi] == '/' {
                    return false;
                }
                vi += 1;
                pi += 1;
            }
            '[' => {
                let Some(end) = p[pi..].iter().position(|&c| c == ']') else {
                    // Unterminated class: literal '[' comparison.
                    if v.get(vi) == Some(&'[') {
                        vi += 1;
                        pi += 1;
                        continue;
                    }
                    return false;
                };
                let end = pi + end;
                let vi_char = match v.get(vi) {
                    Some(&c) => c,
                    None => return false,
                };
                if vi_char == '/' {
                    return false;
                }
                let (negated, class) = match p.get(pi + 1) {
                    Some('!') | Some('^') => (true, &p[pi + 2..end]),
                    _ => (false, &p[pi + 1..end]),
                };
                let mut matched = false;
                let mut ci = 0;
                while ci < class.len() {
                    if ci + 2 < class.len() && class[ci + 1] == '-' {
                        if vi_char >= class[ci] && vi_char <= class[ci + 2] {
                            matched = true;
                        }
                        ci += 3;
                    } else {
                        if vi_char == class[ci] {
                            matched = true;
                        }
                        ci += 1;
                    }
                }
                if matched == negated {
                    return false;
                }
                vi += 1;
                pi = end + 1;
            }
            c => {
                if v.get(vi) != Some(&c) {
                    return false;
                }
                vi += 1;
                pi += 1;
            }
        }
    }
    vi == v.len()
}

/// Does a rule match a specific tool call?
pub fn rule_matches(rule: &Rule, tool_name: &str, args: &Value) -> bool {
    if !tool_name_matches(&rule.tool, tool_name) {
        return false;
    }
    let Some(pattern) = &rule.arg_pattern else {
        return true;
    };
    match primary_arg(tool_name, args) {
        Some(arg) => arg_pattern_matches(pattern, &arg),
        None => false,
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RuleOutcome {
    Deny,
    Allow,
}

/// Evaluate rules. Deny wins; else allow; else `None`.
pub fn evaluate_rules(rules: &RuleSet, tool_name: &str, args: &Value) -> Option<RuleOutcome> {
    for r in &rules.deny {
        if rule_matches(r, tool_name, args) {
            return Some(RuleOutcome::Deny);
        }
    }
    for r in &rules.allow {
        if rule_matches(r, tool_name, args) {
            return Some(RuleOutcome::Allow);
        }
    }
    None
}

/// Merge rule sets — session rules first (added this turn take precedence
/// in evaluation order), then file rules, per the TS wrapper's merge.
pub fn merge_rules(session: &RuleSet, file: &RuleSet) -> RuleSet {
    RuleSet {
        allow: session
            .allow
            .iter()
            .chain(file.allow.iter())
            .cloned()
            .collect(),
        deny: session
            .deny
            .iter()
            .chain(file.deny.iter())
            .cloned()
            .collect(),
    }
}

/// Derive an "Always Allow" rule spec from an approved call: smart globs
/// for bash (`npx:*`, package-manager heads) and file tools (dir/*).
pub fn derive_rule_spec(tool_name: &str, args: &Value) -> String {
    let Some(arg) = primary_arg(tool_name, args) else {
        return tool_name.to_string();
    };

    if tool_name == "bash" {
        let tokens: Vec<&str> = arg.split_whitespace().collect();
        if let Some(first) = tokens.first() {
            if *first == "npx" {
                // "npx package-name" → "bash(npx package-name:*)"
                let pkg: Vec<&str> = tokens.iter().take(2).copied().collect();
                return format!("{tool_name}({}:*)", pkg.join(" "));
            }
            if matches!(*first, "npm" | "yarn" | "pnpm" | "bun" | "deno") && tokens.len() >= 2 {
                let head: Vec<&str> = tokens.iter().take(2).copied().collect();
                return format!("{tool_name}({})", head.join(" "));
            }
        }
        let head = arg.split_whitespace().next().unwrap_or("");
        return format!("{tool_name}({head})");
    }

    if let Some(idx) = arg.rfind('/') {
        let dir = &arg[..idx];
        return format!("{tool_name}({dir}/*)");
    }

    format!("{tool_name}({arg})")
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn parse_rule_bare_and_with_arg() {
        assert_eq!(
            parse_rule("bash"),
            Some(Rule { tool: "bash".into(), arg_pattern: None })
        );
        assert_eq!(
            parse_rule("bash(npm test)"),
            Some(Rule {
                tool: "bash".into(),
                arg_pattern: Some("npm test".into())
            })
        );
        assert_eq!(parse_rule("read_file()").unwrap().arg_pattern, None);
        assert_eq!(parse_rule("  edit:* ( src/* ) ").unwrap().tool, "edit:*");
        assert_eq!(parse_rule(""), None);
        assert_eq!(parse_rule("(x)"), None);
    }

    #[test]
    fn primary_arg_resolution() {
        assert_eq!(primary_arg("bash", &json!({"command": "ls"})), Some("ls".into()));
        assert_eq!(
            primary_arg("git", &json!({"args": ["status", "--short"]})),
            Some("status --short".into())
        );
        assert_eq!(
            primary_arg("read_file", &json!({"path": "src/a.ts"})),
            Some("src/a.ts".into())
        );
        assert_eq!(
            primary_arg("grep", &json!({"pattern": "foo"})),
            Some("foo".into())
        );
        assert_eq!(primary_arg("bash", &json!({})), None);
        assert_eq!(primary_arg("unknown_tool", &json!({"x": 1})), None);
    }

    #[test]
    fn tool_name_pattern_matching() {
        assert!(tool_name_matches("bash", "bash"));
        assert!(tool_name_matches("BASH", "bash"));
        assert!(tool_name_matches("*", "anything"));
        assert!(tool_name_matches("edit:*", "edit_file"));
        assert!(!tool_name_matches("edit:*", "write_file"));
        assert!(!tool_name_matches("bas", "bash"));
    }

    #[test]
    fn arg_pattern_prefix_and_glob() {
        assert!(arg_pattern_matches("npm test", "npm test --watch"));
        assert!(!arg_pattern_matches("npm test", "npmx test"));
        // Quirk kept from the TS (minimatch): ':' is a LITERAL in glob
        // patterns, so the derived "npx cowsay:*" spec matches commands
        // containing that literal colon — never real "npx cowsay -f x"
        // invocations. An allow rule that never fires is the safe direction.
        assert!(!arg_pattern_matches("npx cowsay:*", "npx cowsay -f x"));
        assert!(arg_pattern_matches("npx cowsay:*", "npx cowsay:extra"));
        // `*` does not cross path separators (minimatch semantics).
        assert!(!arg_pattern_matches("src/*", "src/a/b.ts"));
        assert!(arg_pattern_matches("src/*", "src/a.ts"));
        assert!(arg_pattern_matches("*.test.ts", "a.test.ts"));
        assert!(arg_pattern_matches("[a-c]?", "ax"));
        assert!(!arg_pattern_matches("[a-c]?", "dx"));
    }

    #[test]
    fn evaluate_rules_deny_wins_over_allow() {
        let rules = RuleSet {
            allow: vec![parse_rule("bash").unwrap()],
            deny: vec![parse_rule("bash(rm *)").unwrap()],
        };
        assert_eq!(
            evaluate_rules(&rules, "bash", &json!({"command": "rm x"})),
            Some(RuleOutcome::Deny)
        );
        assert_eq!(
            evaluate_rules(&rules, "bash", &json!({"command": "ls"})),
            Some(RuleOutcome::Allow)
        );
        assert_eq!(evaluate_rules(&rules, "grep", &json!({"pattern": "x"})), None);
    }

    #[test]
    fn derive_rule_spec_smart_globs() {
        assert_eq!(
            derive_rule_spec("bash", &json!({"command": "npx cowsay -f moo"})),
            "bash(npx cowsay:*)"
        );
        assert_eq!(
            derive_rule_spec("bash", &json!({"command": "pnpm install --filter pkg"})),
            "bash(pnpm install)"
        );
        assert_eq!(
            derive_rule_spec("bash", &json!({"command": "cargo build --release"})),
            "bash(cargo)"
        );
        assert_eq!(
            derive_rule_spec("edit_file", &json!({"path": "src/lib/a.ts"})),
            "edit_file(src/lib/*)"
        );
        assert_eq!(
            derive_rule_spec("read_file", &json!({"path": "README.md"})),
            "read_file(README.md)"
        );
        assert_eq!(derive_rule_spec("bash", &json!({})), "bash");
    }

    #[test]
    fn loads_rules_from_agents_settings_json() {
        let tmp = tempfile::tempdir().unwrap();
        let file = tmp.path().join(".agents").join("settings.json");
        std::fs::create_dir_all(file.parent().unwrap()).unwrap();
        std::fs::write(
            &file,
            r#"{"permissions": {"allow": ["bash(pnpm test)"], "deny": ["bash(sudo*)"]}}"#,
        )
        .unwrap();
        let rules = load_permission_rules(tmp.path());
        assert_eq!(rules.allow.len(), 1);
        assert_eq!(rules.deny.len(), 1);

        assert_eq!(
            evaluate_rules(&rules, "bash", &json!({"command": "sudo rm x"})),
            Some(RuleOutcome::Deny)
        );
        assert_eq!(
            evaluate_rules(&rules, "bash", &json!({"command": "pnpm test --ci"})),
            Some(RuleOutcome::Allow)
        );
    }

    #[test]
    fn missing_or_malformed_file_yields_empty() {
        let tmp = tempfile::tempdir().unwrap();
        assert_eq!(load_permission_rules(tmp.path()), RuleSet::default());
        let file = tmp.path().join(".agents").join("settings.json");
        std::fs::create_dir_all(file.parent().unwrap()).unwrap();
        std::fs::write(&file, "not json").unwrap();
        assert_eq!(load_permission_rules(tmp.path()), RuleSet::default());
        std::fs::write(&file, r#"{"permissions": "nope"}"#).unwrap();
        assert_eq!(load_permission_rules(tmp.path()), RuleSet::default());
    }

    #[test]
    fn merge_puts_session_rules_first() {
        let session = RuleSet {
            allow: vec![Rule { tool: "bash".into(), arg_pattern: None }],
            deny: vec![],
        };
        let file = RuleSet {
            allow: vec![Rule { tool: "grep".into(), arg_pattern: None }],
            deny: vec![],
        };
        let merged = merge_rules(&session, &file);
        assert_eq!(merged.allow.len(), 2);
        assert_eq!(merged.allow[0].tool, "bash");
    }
}
